src/viewers/VivViewer.js

```diff
@@ -0,0 +1,164 @@
+    // Save the view state and trigger rerender.
+    const { views } = this.props;
+    this.setState(prevState => {
+      const viewStates = {};
+      views.forEach(view => {
+        const currentViewState = prevState.viewStates[view.id];
+        viewStates[view.id] = view.filterViewState({
+          viewState: { ...viewState, id: viewId },
+          oldViewState,
+          currentViewState
+        });
+      });
+      return { viewStates };
+    });
+  }
+
+  /**
+   * This updates the viewStates' height and width with the newest height and
+   * width on any call where the viewStates changes (i.e resize events),
+   * using the previous state (falling back on the view's initial state) for target x and y, zoom level etc.
+   */
+  componentDidUpdate(prevProps) {
+    const { views } = this.props;
+    // Update internal viewState on view changes as well as height and width changes.
+    // Maybe we should add x/y too?
+    if (
+      views.some(
+        (view, i) =>
+          prevProps.views[i] !== view ||
+          view.initialViewState.height !==
+            prevProps.views[i].initialViewState.height ||
+          view.initialViewState.width !==
+            prevProps.views[i].initialViewState.width
+      )
+    ) {
+      // eslint-disable-next-line react/no-did-update-set-state
+      this.setState(prevState => {
+        const viewStates = {};
+        views.forEach(view => {
+          const { height, width } = view.initialViewState;
+          const currentViewState = prevState.viewStates[view.id];
+          viewStates[view.id] = view.filterViewState({
+            viewState: {
+              ...(currentViewState || view.initialViewState),
+              height,
+              width,
+              id: view.id
+            }
+          });
+        });
+        return { viewStates };
+      });
+    }
+  }
+
+  /**
+   * This renders the layers in the DeckGL context.
+   */
+  _renderLayers() {
+    const { viewStates } = this.state;
+    const { views, layerProps } = this.props;
+    return views.map((view, i) =>
+      view.getLayers({ viewStates, props: layerProps[i] })
+    );
+  }
+
+  render() {
+    /* eslint-disable react/destructuring-assignment */
+    const { views, randomize } = this.props;
+    const { viewStates } = this.state;
+    const deckGLViews = views.map(view => view.getDeckGlView());
+    // DeckGL seems to use the first view more than the second for updates
+    // so this forces it to use the others more evenly.  This isn't perfect,
+    // but I am not sure what else to do.  The DeckGL render hooks don't help,
+    // but maybe useEffect() would help?  I couldn't work it out as
+    // The issue is that I'm not sure how React would distinguish between forced updates
+    // from permuting the views array and "real" updates like zoom/pan.
+    // I tried keeping a counter but I couldn't figure out resetting it
+    // without triggering a re-render.
```

Comment:
I would look into `React.memo`. https://blog.bitsrc.io/improve-performance-in-react-functional-components-using-react-memo-b2e80c11e15a. 

Effectively you memoize the return of a functional component, but also supply a comparator which compares prevProps and nextProps.

```javascript
const App = (props) => { /* ...some expensive logic */ return <div>Calculated expensive thing</div> } 
const shouldUpdate = (prevProps, nextProps) => {
  return prevProps.some.nested.thing !== nextProps.some.nested.thing;
}
export default React.memo(App, shouldUpdate);
```

Reply:
Hmmm...Ok, will do.

Reply:
See my other comment.  Feel free to respond!